lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -164,12 +174,35 @@ void FIRRTLTypesLowering::runOnOperation() {
   while (!opsToRemove.empty())
     opsToRemove.pop_back_val()->erase();
 
-  // Remove args that have been lowered.
-  getOperation().eraseArguments(argsToRemove);
+  if (argsToRemove.empty())
+    return;
+
+  // Remove block args that have been lowered.
+  body->eraseArguments(argsToRemove);
   argsToRemove.clear();
 
+  // Remember the original argument attributess.
+  SmallVector<NamedAttribute, 8> originalArgAttrs;
+  DictionaryAttr originalAttrs = module->getAttrDictionary();
+  for (size_t i = 0, e = originalNumArgs; i < e; ++i)
+    originalArgAttrs.push_back(
+        originalAttrs.getNamed(getArgAttrName(i)).getValue());
+  DictionaryAttr::sortInPlace(originalArgAttrs);
+
+  // Copy over any attributes that weren't original argument attributes.
+  auto *argAttrBegin = originalArgAttrs.begin();
+  auto *argAttrEnd = originalArgAttrs.end();
+  for (auto attr : originalAttrs)
+    if (std::lower_bound(argAttrBegin, argAttrEnd, attr) == argAttrEnd)
+      newAttrs.push_back(attr);
```

Comment:
This is looking good, but I think it can still be made better.  Feel free to land this and we can keep improving in place though!

This can be fundamentally a single pass algorithm, something like this:

```
for (auto attr : module->getAttrDictionary()) {
   // If this is the name of an argument attribute, return the number for it.
  Optional<unsigned> argNumber = getArgAttrNumber(attr.name);
  if (!argNumber.hasValue() || weAreKeepingThisArgument(argNumber.get())) 
     newAttrs.push_back(attr);
}
```
It requires introducing a "getArgAttrNumber" sort of helper function (which should be generally useful anyway, akin to `getArgAttrName`) and a way to know which argument ID's are to be deleted (which could be stored in a SmallSet next to argsToRemove.

Reply:
Yep, I have had that thought, but wasn't sure if defining `getArgAttrNumber(Identifier)` was going to be too hacky/stringly-typed (see comment [here](https://github.com/llvm/circt/pull/651#issuecomment-786296652)).

Reply:
Thanks.  I think the arg name attribute design is inherently stringly typed, so embracing that with an "atoi" makes sense.   StringRef has a helper for this.